src/main/io/gps.c

```diff
@@ -1512,185 +2075,244 @@ typedef enum {
-//15:17:55  R -> UBX 03-09,  Size 208,  'Unknown'
-//15:17:55  R -> UBX 03-10,  Size 336,  'Unknown'
-//15:17:55  R -> UBX NAV-SOL,  Size  60,  'Navigation Solution'
-//15:17:55  R -> UBX NAV,  Size 100,  'Navigation'
-//15:17:55  R -> UBX NAV-SVINFO,  Size 328,  'Satellite Status and Information'
-
-// from the UBlox9 document, the largest payout we receive is the NAV-SAT and the payload size
-// is calculated as 8 + 12*numCh.  numCh in the case of a M9N is 42.
-#define UBLOX_PAYLOAD_SIZE (8 + 12 * 42)
-
-
-// Receive buffer
+// Do we have a new valid fix?
+static bool ubxHaveNewValidFix = false;
+
+// Do we have new position information?
+static bool ubxHaveNewPosition = false;
+
+// Do we have new speed information?
+static bool ubxHaveNewSpeed = false;
+
+// From the UBX protocol docs, the largest payload we receive is NAV-SAT, which
+// is calculated as 8 + 12*numCh. Max reported sats can be up to 56.
+// We're using the max for M8 (32) for our sizing, since Configurator only
+// supports a max of 32 sats and we want to limit the payload buffer space used.
+#define UBLOX_PAYLOAD_SIZE (8 + 12 * GPS_SV_MAXSATS_M8N)
+#define UBLOX_MAX_PAYLOAD_SANITY_SIZE 776   // Any returned payload length greater than a 64 sat NAV-SAT is considered unreasonable, and probably corrupted data.
+
+// Received message frame fields.
+// - Preamble sync character 1 & 2 are not saved, only detected for parsing.
+// - Message class & message ID indicate the type of message receieved.
+static uint8_t ubxRcvMsgClass;
+static uint8_t ubxRcvMsgID;
+// - Payload length assembled from the length LSB & MSB bytes.
+static uint16_t ubxRcvMsgPayloadLength;
+// - Payload, each message type has its own payload field layout, represented by the elements of this union.
+//   Note that the size of the buffer is determined by the longest possible payload, currently UBX-NAV-SAT.
+//   See size define comments above. Warning, this is fragile! If another message type becomes the largest
+//   payload instead of UBX-NAV-SAT, UBLOX_PAYLOAD_SIZE above needs to be adjusted!
 static union {
-    ubxNavPosllh_t posllh;
-    ubxNavStatus_t status;
-    ubxNavDop_t dop;
-    ubxNavSolution_t solution;
-    ubxNavVelned_t velned;
-    ubxNavPvt_t pvt;
-    ubxNavSvinfo_t svinfo;
-    ubxNavSat_t sat;
-    ubxCfgGnss_t gnss;
-    ubxAck_t ack;
-    uint8_t bytes[UBLOX_PAYLOAD_SIZE];
-} _buffer;
-
-void _update_checksum(uint8_t *data, uint8_t len, uint8_t *ck_a, uint8_t *ck_b)
+    ubxNavPosllh_t ubxNavPosllh;
+    ubxNavStatus_t ubxNavStatus;
+    ubxNavDop_t ubxNavDop;
+    ubxNavSol_t ubxNavSol;
+    ubxNavVelned_t ubxNavVelned;
+    ubxNavPvt_t ubxNavPvt;
+    ubxNavSvinfo_t ubxNavSvinfo;
+    ubxNavSat_t ubxNavSat;
+    ubxCfgGnss_t ubxCfgGnss;
+    ubxMonVer_t ubxMonVer;
+    ubxAck_t ubxAck;
+    uint8_t rawBytes[UBLOX_PAYLOAD_SIZE];  // Used for adding raw bytes to the payload. WARNING: This byte array must be as large as the largest payload for any message type above!
+} ubxRcvMsgPayload;
```

Comment:
This will cause large changeset. Is rename really necessary now?

Reply:
We had complaints about the code being difficult to read and review.  And while attempting to resolve issues, it really helps to have consistent and intuitive naming structures.  When this PR is merged, it will be easier to move forward if we get this done now.

Reply:
Agree. Normally should be done in subsequent refactoring PR. But this weekend we are in feature freeze. So let's do it now as it would not be allowed in a patch.